langchain/src/util/websocket_stream.ts

```diff
@@ -0,0 +1,101 @@
+import WsWebSocket from "ws";
+import { getEnv } from "./env.js";
+
+export interface WebSocketConnection<
+  T extends Uint8Array | string = Uint8Array | string
+> {
+  readable: ReadableStream<T>;
+  writable: WritableStream<T>;
+  protocol: string;
+  extensions: string;
+}
+
+export interface WebSocketCloseInfo {
+  code?: number;
+  reason?: string;
+}
+
+export interface WebSocketStreamOptions {
+  protocols?: string[];
+  signal?: AbortSignal;
+}
+
+/**
+ * [WebSocket](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket) with [Streams API](https://developer.mozilla.org/en-US/docs/Web/API/Streams_API)
+ *
+ * @see https://web.dev/websocketstream/
+ */
+export class WebSocketStream<
+  T extends Uint8Array | string = Uint8Array | string
+> {
+  readonly url: string;
+
+  readonly connection: Promise<WebSocketConnection<T>>;
+
+  readonly closed: Promise<WebSocketCloseInfo>;
+
+  readonly close: (closeInfo?: WebSocketCloseInfo) => void;
+
+  constructor(url: string, options: WebSocketStreamOptions = {}) {
+    if (options.signal?.aborted) {
+      throw new DOMException("This operation was aborted", "AbortError");
+    }
+
+    this.url = url;
+
+    let ws: WebSocket;
+
+    if (getEnv() === "browser") {
+      ws = new WebSocket(url, options.protocols ?? []);
```

Comment:
I unfortunately think even the presence of the import statement will break most bundlers - might be nice to find some compatible package? Or just mark this as `Node` only

Reply:
We can also have separate entrypoints for web vs. Node and factor out common logic:

https://github.com/langchain-ai/langchainjs/tree/main/langchain/src/chat_models/googlevertexai

Reply:
yes, i make two entry for xinghuo chat model in new commit. It's a good idea.